syntax = "proto3";

package catalogue;

service CatalogueService {
    rpc GetProduct(GetProductRequest) returns (ProductResponse);
    rpc CreateProduct(CreateProductRequest) returns (ProductResponse);
    rpc UpdateProduct(UpdateProductRequest) returns (ProductResponse);
    rpc DeleteProduct(DeleteProductRequest) returns (Empty);
    rpc ListProducts(Empty) returns (ListProductsResponse);

    // Category CRUD operations
    rpc GetCategory(GetCategoryRequest) returns (CategoryResponse);
    rpc CreateCategory(CreateCategoryRequest) returns (CategoryResponse);
    rpc UpdateCategory(UpdateCategoryRequest) returns (CategoryResponse);
    rpc DeleteCategory(DeleteCategoryRequest) returns (Empty);
    rpc ListCategories(Empty) returns (ListCategoriesResponse);

    // Product Variant CRUD operations
    rpc GetProductVariant(GetProductVariantRequest) returns (ProductVariantResponse);
    rpc CreateProductVariant(CreateProductVariantRequest) returns (ProductVariantResponse);
    rpc UpdateProductVariant(UpdateProductVariantRequest) returns (ProductVariantResponse);
    rpc DeleteProductVariant(DeleteProductVariantRequest) returns (Empty);
    rpc ListProductVariants(ListProductVariantsRequest) returns (ListProductVariantsResponse);
}

message GetProductRequest {
    string hash = 1;
}

message CreateProductRequest {
    string name = 1;
    string description = 2;
    string category_hash = 3;
    string brand_hash = 4;
    double base_price = 5;
    int32 stock = 6;
    string sku = 7;
    double weight = 8;
    string dimensions = 9;
    bool is_active = 10;
    repeated ProductImage images = 11;
    repeated ProductAttribute attributes = 12;
    repeated ProductVariant variants = 13;
}

message UpdateProductRequest {
    string hash = 1;
    string name = 2;
    string description = 3;
    string category_hash = 4;
    string brand_hash = 5;
    double base_price = 6;
    int32 stock = 7;
    string sku = 8;
    double weight = 9;
    string dimensions = 10;
    bool is_active = 11;
    repeated ProductImage images = 12;
    repeated ProductAttribute attributes = 13;
    repeated ProductVariant variants = 14;
}

message DeleteProductRequest {
    string hash = 1;
}

message ProductImage {
    string image_url = 1;
    string alt_text = 2;
    bool is_primary = 3;
}

message ProductAttribute {
    string name = 1;
    string value = 2;
}

message ProductVariant {
    string name = 1;
    map<string, string> attributes = 2;
    double price = 3;
    int32 stock = 4;
    string sku = 5;
    string image_url = 6;
    string hash = 7;
}

message ProductResponse {
    string hash = 1;
    string name = 2;
    string description = 3;
    string category_hash = 4;
    string brand_hash = 5;
    double base_price = 6;
    int32 stock = 7;
    string sku = 8;
    double weight = 9;
    string dimensions = 10;
    bool is_active = 11;
    repeated ProductImage images = 12;
    repeated ProductAttribute attributes = 13;
    repeated ProductVariant variants = 14;
}

message ListProductsResponse {
    repeated ProductResponse products = 1;
}

// Category messages
message GetCategoryRequest {
    string hash = 1;
}

message CreateCategoryRequest {
    string name = 1;
    string description = 2;
    string parent_hash = 3;
}

message UpdateCategoryRequest {
    string hash = 1;
    string name = 2;
    string description = 3;
    string parent_hash = 4;
}

message DeleteCategoryRequest {
    string hash = 1;
}

message CategoryResponse {
    string hash = 1;
    string name = 2;
    string description = 3;
    string parent_hash = 4;
}

message ListCategoriesResponse {
    repeated CategoryResponse categories = 1;
}

// Product Variant messages
message GetProductVariantRequest {
    string hash = 1;
}

message CreateProductVariantRequest {
    string product_hash = 1;
    string name = 2;
    map<string, string> attributes = 3;
    double price = 4;
    int32 stock = 5;
    string sku = 6;
    string image_url = 7;
    string hash = 8;
}

message UpdateProductVariantRequest {
    string hash = 1;
    string name = 2;
    map<string, string> attributes = 3;
    double price = 4;
    int32 stock = 5;
    string sku = 6;
    string image_url = 7;
}

message DeleteProductVariantRequest {
    string hash = 1;
}

message ProductVariantResponse {
    string hash = 1;
    string product_hash = 2;
    string name = 3;
    map<string, string> attributes = 4;
    double price = 5;
    int32 stock = 6;
    string sku = 7;
    string image_url = 8;
}

message ListProductVariantsRequest {
    string product_hash = 1;
}

message ListProductVariantsResponse {
    repeated ProductVariantResponse variants = 1;
}

message Empty {}
